### PR TITLE
CFE-977,CFE-980: Introduce Archiver interface and its implementation MirrorArchiver

### PR DESCRIPTION
**Pull request by sherine-k**
# Description
BuildArchive creates an archive that contains:
* docker/v2/repositories : manifests for all mirrored images
* docker/v2/blobs/sha256 : blobs that haven't been mirrored (diff)
* working-dir
* hidden history file
* image set config

This PR updates the way digests are processed. Use of github.com/opencontainers/go-digest's ParseDigest is prefered to splitting the digest on `@sha256:` so that no assumptions are made with regards to the hashing algorithm.

Fixes # [CFE-980](https://issues.redhat.com/browse/CFE-980)

TODO: add calls to history package. Depends on  # [CFE-978](https://issues.redhat.com/browse/CFE-978)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Unit test archive_test.go
This interface is not yet plugged into executor.go, but should be later, with [CFE-991](https://issues.redhat.com/browse/CFE-991)

## Expected Outcome

Unit tests should pass

### CODE REVIEW

**Review by lmzuccarelli — COMMENTED** [review]
@sherine-k -  could you add in the comments section how to verify/test this feature i.e build a simple archive and  verify its contents ?

File: v2/pkg/image/image.go [hunk 1]
```diff
@@ -59,7 +62,12 @@ func ParseRef(imgRef string) (ImageSpec, error) {
 	if strings.Contains(imgSpec.Name, "@") {
 		imgSplit := strings.Split(imgSpec.Name, "@")
 		if len(imgSplit) > 1 {
-			imgSpec.Digest = strings.Split(imgSplit[1], ":")[1]
+			validDigest, err := digest.Parse(imgSplit[1])
+			if err != nil {
+				return ImageSpec{}, fmt.Errorf("unable to parse image %s correctly, invalid digest: %v", imgRef, err)
+			}
+			imgSpec.Digest = validDigest.Hex()
```
Comment (aguidirh) [inline v2/pkg/image/image.go:69]: Hi @sherine-k,

It seems the Hex() func is deprecated. It is saying to use Digest.Encoded instead

File: v2/pkg/manifest/oci-manifest.go [hunk 2]
```diff
@@ -140,10 +142,11 @@ func (o *Manifest) GetRelatedImagesFromCatalogByFilter(filePath, label string, o
 func (o *Manifest) ExtractLayersOCI(fromPath, toPath, label string, oci *v1alpha3.OCISchema) error {
 	if _, err := os.Stat(toPath + "/" + label); errors.Is(err, os.ErrNotExist) {
 		for _, blob := range oci.Layers {
-			if !strings.Contains(blob.Digest, "sha256") {
+			validDigest, err := digest.Parse(blob.Digest)
+			if err != nil {
 				return fmt.Errorf("the digest format is not correct %s ", blob.Digest)
 			}
-			f, err := os.Open(fromPath + "/" + strings.Split(blob.Digest, ":")[1])
+			f, err := os.Open(fromPath + "/" + validDigest.Hex())
```
Comment (aguidirh) [inline v2/pkg/manifest/oci-manifest.go:149]: Hi @sherine-k,

It seems the Hex() func is deprecated. It is saying to use Digest.Encoded instead

File: v2/pkg/operator/collector.go [hunk 3]
```diff
@@ -99,16 +101,16 @@ func (o *Collector) OperatorImageCollector(ctx context.Context) ([]v1alpha3.Copy
...
-					return []v1alpha3.CopyImageSchema{}, fmt.Errorf("[OperatorImageCollector] the disgets seems to incorrect for %s ", op.Catalog)
-				}
 			}
-			manifest := strings.Split(oci.Manifests[0].Digest, ":")[1]
+			validDigest, err := digest.Parse(oci.Manifests[0].Digest)
+			if err != nil {
+				return []v1alpha3.CopyImageSchema{}, fmt.Errorf("[OperatorImageCollector] the disgets seems to be incorrect for %s ", op.Catalog)
+			}
+
+			manifest := validDigest.Hex()
```
Comment (aguidirh) [inline v2/pkg/operator/collector.go:113]: Hi @sherine-k,

It seems the Hex() func is deprecated. It is saying to use Digest.Encoded instead

File: v2/pkg/operator/local_stored_collector.go [hunk 4]
```diff
@@ -98,24 +101,28 @@ func (o *LocalStorageCollector) OperatorImageCollector(ctx context.Context) ([]v
...
-			}
 		}
-		manifest := strings.Split(oci.Manifests[0].Digest, ":")[1]
-		o.Log.Info("manifest %v", manifest)
+		validDigest, err := digest.Parse(oci.Manifests[0].Digest)
+		if err != nil {
+			return []v1alpha3.CopyImageSchema{}, fmt.Errorf("[OperatorImageCollector] the disgests seem to incorrect for %s: %v ", op.Catalog, err)
+		}
 
+		manifest := validDigest.Hex()
```
Comment (aguidirh) [inline v2/pkg/operator/local_stored_collector.go:110]: Hi @sherine-k,

It seems the Hex() func is deprecated. It is saying to use Digest.Encoded instead

File: v2/pkg/release/local_stored_collector.go [hunk 5]
```diff
@@ -105,10 +107,15 @@ func (o *LocalStorageCollector) ReleaseImageCollector(ctx context.Context) ([]v1
 			if len(oci.Manifests) == 0 {
 				return []v1alpha3.CopyImageSchema{}, fmt.Errorf(errMsg, "image index not found ")
 			}
-			manifest := strings.Split(oci.Manifests[0].Digest, ":")[1]
+			validDigest, err := digest.Parse(oci.Manifests[0].Digest)
+			if err != nil {
+				return []v1alpha3.CopyImageSchema{}, fmt.Errorf("[ReleaseImageCollector] invalid digest for image index %s: %v", oci.Manifests[0].Digest, err)
+			}
+
+			manifest := validDigest.Hex()
```
Comment (aguidirh) [inline v2/pkg/release/local_stored_collector.go:115]: Hi @sherine-k,

It seems the Hex() func is deprecated. It is saying to use Digest.Encoded instead

**Pull request merged**